### ISSUE TITLE
`ttnn.experimental.minimal_matmul` has inconsistent pcc when acting on the same input rows

### ISSUE THREAD

**Issue by rdraskicTT** [tenstorrent/tt-metal#35167]
`ttnn.experimental.minimal_matmul` is creating differing outputs for identical users in batched prefill for llama 70B. We need user logits to be the same for same inputs in order to have deterministic execution.

Unit test repro can be found at `rdraskic/minimal-matmul-bug` and can be run with `pytest test_minimal_matmul.py`. Switching to `ttnn.linear` makes the test pass and is currently applied as a workaround.

**Comment by cglagovichTT** [comment]
Ack, I'm looking at this. Was able to make a single chip repro.

**Comment by cglagovichTT** [comment]
## Summary
When using ttnn.experimental.minimal_matmul with identical input activation rows (e.g., created via torch.Tensor.expand), the output rows are not identical. This is unexpected since identical inputs should produce identical outputs.
## Root Cause
The k_forward serpentine pattern (used for K-block ordering optimization) is not reset at the start of each M-block. Since different cores may have different numbers of N-blocks assigned, the k_forward state diverges across cores after the first M-block:
Cores with even N-blocks: k_forward = true
Cores with odd N-blocks: k_forward = false
This causes K-blocks to be accumulated in different orders for different M-blocks. While mathematically equivalent, **floating-point non-associativity** causes small numerical differences in outputs that should be identical.
## Affected Configuration
FP16 accumulation (fp32_dest_acc_en=False)
Multiple M-blocks per core with non-uniform N-block distribution
## Workaround
Enable FP32 accumulation (fp32_dest_acc_en=True) to reduce floating-point precision errors. This resolves the issue for bfloat8_b and bfloat16 inputs.
## Proposed Fix
Add a configuration flag to MinimalMatmulConfig that ensures consistent K-block ordering across all M-blocks (by resetting k_forward at the start of each M-block). This would trade some performance optimization for numerical consistency when required.

@rdraskicTT now that we know it's a floating point associativity issue, would you like to go forward with the proposed fix? It will lead to slightly degraded performance.
Your alternative is to apply the workaround of fp32 accumulation, though it's not foolproof and you still might find some cases where users have different outputs due to non-associativity.

**Comment by cglagovichTT** [comment]
I took a look at performance of minimal matmul on main and with in1 reuse disabled, which would resolve the issue. There is a very small performance delta between the two so I think it makes sense to proceed. @rdraskicTT thoughts?

baseline
```
DTYPE: dtype_bf16, FP32 ACC: fp32_acc, FIDELITY: HiFi2
| M, K, N | math util (%) | measured perf (ms) | attributes |
|---|---:|---:|---:|
| (512, 512, 512) | 11.2 | 0.018 | 'MinimalMatmulConfig(M_block_size=2;K_block_size=2;N_block_size=2;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (512, 1024, 1024) | 19.4 | 0.042 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=4;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (512, 1024, 2048) | 23.0 | 0.071 | 'MinimalMatmulConfig(M_block_size=2;K_block_size=4;N_block_size=4;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 1024, 1024) | 26.5 | 0.062 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=2;N_block_size=4;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 1024, 2048) | 34.4 | 0.095 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 2048, 2048) | 43.1 | 0.152 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 2048, 2048) | 45.5 | 0.288 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 2048, 3072) | 46.3 | 0.425 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 3072, 3072) | 51.2 | 0.577 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 3072, 3072) | 53.7 | 0.824 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 3072, 4096) | 55.0 | 1.072 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 4096, 4096) | 57.9 | 1.358 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (4096, 4096, 4096) | 58.7 | 1.786 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (8192, 8192, 8192) | 63.7 | 13.167 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (16384, 16384, 16384) | 64.8 | 103.603 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
```

No in1 reuse
```
DTYPE: dtype_bf16, FP32 ACC: fp32_acc, FIDELITY: HiFi2
| M, K, N | math util (%) | measured perf (ms) | attributes |
|---|---:|---:|---:|
| (512, 512, 512) | 11.2 | 0.018 | 'MinimalMatmulConfig(M_block_size=2;K_block_size=2;N_block_size=2;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (512, 1024, 1024) | 19.4 | 0.042 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=4;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (512, 1024, 2048) | 22.1 | 0.074 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 1024, 1024) | 26.6 | 0.061 | 'MinimalMatmulConfig(M_block_size=16;K_block_size=2;N_block_size=4;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 1024, 2048) | 34.3 | 0.096 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 2048, 2048) | 42.6 | 0.154 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 2048, 2048) | 45.9 | 0.286 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 2048, 3072) | 47.3 | 0.416 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 3072, 3072) | 50.9 | 0.579 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 3072, 3072) | 53.4 | 0.828 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 3072, 4096) | 53.9 | 1.094 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 4096, 4096) | 57.8 | 1.361 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (4096, 4096, 4096) | 59.2 | 1.772 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (8192, 8192, 8192) | 63.8 | 13.151 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (16384, 16384, 16384) | 64.8 | 103.484 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
```

**Comment by rdraskicTT** [comment]
@cglagovichTT thanks for looking into this! I agree that this perf drop seems negligible so we can probably do it that way. How to disable in1, will you make a PR for this?

**Comment by cglagovichTT** [comment]
yes I'll open a PR to disable in1 reuse

**Comment by cglagovichTT** [comment]
Recollected perf with no in1 reuse and row-major output block ordering.
Perf is confirmed basically same as before, going forward with removing serpentine schedule.

```
DTYPE: dtype_bf16, FP32 ACC: fp32_acc, FIDELITY: HiFi2
| M, K, N | math util (%) | measured perf (ms) | attributes |
|---|---:|---:|---:|
| (512, 512, 512) | 11.2 | 0.018 | 'MinimalMatmulConfig(M_block_size=2;K_block_size=2;N_block_size=2;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (512, 1024, 1024) | 19.4 | 0.042 | 'MinimalMatmulConfig(M_block_size=2;K_block_size=2;N_block_size=4;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (512, 1024, 2048) | 22.9 | 0.071 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=4;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 1024, 1024) | 26.6 | 0.062 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 1024, 2048) | 34.7 | 0.095 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (1024, 2048, 2048) | 43.0 | 0.152 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 2048, 2048) | 45.8 | 0.286 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=4;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 2048, 3072) | 48.4 | 0.406 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=4;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (2048, 3072, 3072) | 51.6 | 0.571 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 3072, 3072) | 53.9 | 0.821 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=16;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 3072, 4096) | 55.6 | 1.061 | 'MinimalMatmulConfig(M_block_size=4;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (3072, 4096, 4096) | 57.8 | 1.360 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (4096, 4096, 4096) | 58.4 | 1.796 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (8192, 8192, 8192) | 62.9 | 13.342 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
| (16384, 16384, 16384) | 64.6 | 103.826 | 'MinimalMatmulConfig(M_block_size=8;K_block_size=8;N_block_size=8;subblock_h=2;subblock_w=2;compute_with_storage_grid_size=(x=8;y=8))';  |
```
